File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        HashMap <Integer,Integer> nextMap = new HashMap<Integer,Integer>();
        Stack <Integer> s = new Stack<Integer>();
        int n = nums2.length;
        // s.push(nums2[n-1]);
        for(int i = n-1;i>= 0;i--){
            while(!s.isEmpty() && nums2[i]>= s.peek()){
                s.pop();
            }
            int nextGreater = s.isEmpty() ? -1 : s.peek();
            nextMap.put(nums2[i],nextGreater);
            s.push(nums2[i]);

        }
        int result [] = new int[nums1.length];
        for (int i = 0;i<nums1.length;i++){
            result[i] = nextMap.get(nums1[i]);
        }

        return result;


        
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        nextGreater = {}
        res = []
        
        for i in range(len(nums2) - 1,-1,-1):
            while stack and nums2[i] > stack[-1]:
                stack.pop()
            if stack:
                nextGreater[nums2[i]] = stack[-1]
            else:
                nextGreater[nums2[i]] = -1
            stack.append(nums2[i])
        
        for num in nums1:
            res.append(nextGreater[num])
        return res

        